### DISCUSSION TITLE
Unable to boot from iSCSI device when using iPXE launched from uBoot

### DISCUSSION THREAD

**Discussion by alan-alanson** [ipxe/ipxe#443] [General]
Hello,

My goal is to boot my Xilinx ZCU102 based device from an iSCSI drive (hosted on a Centos machine using SCST ). Previously I have been using TFTP directly from uboot. 

So far I have built the snp.efi executable with an embedded script that just opens an iPXE shell. I launch iPXE from uBoot using the bootefi command after loading it in to memory from an SD card. I obtain an IP address using DHCP from the centos machine then try to attach an iSCSI device using the sanhook command :

`sanhook –no-describe iscsi:<centos-mahcine-ip>:::0:iqn.<target-name>:target1`

This command returns success and indicates a san device has been registered. Observing packet captures and logs on my iSCSI host machine appears to confirm that there was a successful login to the iSCSI target.

The iSCSI target I try to attach to in this command is a partition on a hard-drive in the centos machine, which has been formatted to FAT-32 W95 and contains a linux kernel image with an efi stub (I have confirmed that I can boot from this directly from uBoot with bootefi when I use TFTP).

I see that the device handle given to the attached device is (when printed using efi_devpath_text)  :

`0x80 has device path /VenHw(<GUID>)/MAC(<mac-address>,1)/ MAC(<mac-address>,1)/UNKNOWN(0003,0013)`

After looking at the UEFI spec and uBoot’s efi api, the 0003 corresponds to a Messaging Device Path type and the sub type (00013) corresponds to an iSCSI device. It seems this definition is missing from uBoot’s efi api.

However, I am unable to boot from the attached SAN device:

`sanboot –no-describe –filename Image.gz`

After turning on debug prints and looking through the iPXE source code, it appears that if my device is not being registered with the Simple Filesystem Protocol when I run sanhook. This seems to be at odds with this slide from [this presentation](https://archive.fosdem.org/2020/schedule/event/firmware_duwu/attachments/slides/3867/export/events/attachments/firmware_duwu/slides/3867/Discover_UEFI_with_U_Boot_16_9.pdf) I have been referring to:

![ipxe](https://user-images.githubusercontent.com/88894536/129386975-346be251-4367-4521-bae2-c66012508a85.png)

I have seen that there is a call to ConnectController (which is handled in uBoots efi api), and afterwards I print out the supported protocols and Simple Filesystem Protocol is not present. Does this indicate that uBoot is not registering the device's protocols correctly? 

Has anyone else had any success doing this? Or could someone give me some debugging tips please? 
Many thanks.

(NB I am using xilinx's v2020-2 version of uBoot)

**Comment by NiKiZe** [comment]
You mentioned partition, have you tried using the disk instead?
What is the exact output you see during this boot failure?
Why are you using `--no-describe` ?

Are you always running sanhook+sanboot and if so why?

**Comment by alan-alanson** [comment]
Thanks for the reply.

I was using `--no-describe` and sanhook + sanboot becasue of this guide [here](https://www.xypron.de/projects/u-boot/iscsi.html).
When I don't use it I get and error saying "EFIBLK has no ACPI table protocol" and "Could not describe SAM devices: Operation not supported http://ipxe.org/3c222083". There is an uBoot config option to enable ACPI table generation but it seems to be only supported on x86.  

If I just run `sanboot –no-describe iscsi:<centos-mahcine-ip>:::0:iqn.<target-name>:target1`

The output says “`SAN 0x80.1 could not initiate read capacity: Error 0x3f122083`” and after “`Could not open SAN device Error 0x3f122083`”. Interestingly, when I run the sanhook command it does not fail to initiate the read capacity command.

I have tried your suggestion of configuring my iSCSI target to point to the entire disk (not just the partition) and that appears to get things further! 

Running  sanhook then sanboot still fails, however I now see it finding child devices, and the device paths show the two partitions I have on the hard disk.

It now fails and prints “`EFIBLK 0x80 could not load image: Error 0x7f22208e`” : Having looked at the corresponding error page and put some debug prints in to uBoot. From the debug prints in uBoot, it seems to fail at an EFI_CALL to open the file I specify with the `–filename` option of sanboot. I tried placing my boot image in /EFI/BOOT/BOOTAA64.EFI on the efi partition of the hard disk, however I still get the same result/ error. After uBoot makes the call to open, I also get 3 prints of "`NETDEV net0 transmission 0x0 failed: Error 0x2a086089 ...`".

**Reply by NiKiZe** [reply]
Could you please post the full output of iPXE?
First of the exact iPXE version,

Did you follow  http://ipxe.org/3c222083 and read that page? you will not be able to boot this unless you get your firmware fixed - but let's start with actually making sure that you are using a recent iPXE version.

You should have more urls when getting errors - follow those urls

**Reply by alan-alanson** [reply]
Here is the output from running sanhook+sanboot commands:

```
iPXE 1.21.1+ (ge09e1) -- Open Source Network Boot Firmware -- http://ipxe.org
Features: DNS HTTP iSCSI TFTP VLAN AoE EFI Menu
IMAGE myscript.ipxe unregistered
iPDXE script started
Opening shell
iPXE> set initiator-iqn iqn.2021-07.ipxe:test-unit-1.test && dhcp
NETDEV net0 opening
SNP net0 could not set station address before initialising: Error 0x7f594083
SNP net0 could not set receive filters 0x00 (have 0x00): Error 0x7f594083 (http://ipxe.org/7f594083)
SNP net0 could not set receive filters 0x07 (have 0x00): Error 0x7f594083 (http://ipxe.org/7f594083)
SNP net0 could not set receive filters 0x15 (have 0x00): Error 0x7f594083 (http://ipxe.org/7f594083)
SNP net0 could not set receive filters 0x05 (have 0x00): Error 0x7f594083 (http://ipxe.org/7f594083)
SNP net0 could not set receive filters 0x01 (have 0x00): Error 0x7f594083 (http://ipxe.org/7f594083)
SNPDEV 0x78e188a8 link is up
NETDEV net0 starting configuration via ipv6
NETDEV net0 starting configuration via dhcp
Configuring (net0 <REDACTED_MAC_ADDRESS>)...... NETDEV net0 configured via dhcp
NETDEV net0 configuration via ipv6 failed: Connection timed out (http://ipxe.org/4c076092)
  ok
iPXE> sanhook --no-describe iscsi:<IP_ADDRESS>:::0:iqn.2021-07.com.test:target1
iSCSI 0x78e19fe0+28 opening iscsi URI
iSCSI 0x78e1a0e8 initiator iqn.2021-07.ipxe:test-unit-1.test
iSCSI 0x78e1a0e8 target iqn.2021-07.com.test:target1
iSCSI 0x78e1a0e8 entering security negotiation
SCSI 0x78e1a388 created for LUN 0000-0000-0000-0000
INTF 0x78e1a0e8+50 opening (SOCK_STREAM,AF_INET) socket
NETDEV net0 transmission 0x0 failed: Error 0x2a086089 (http://ipxe.org/2a086089)
iSCSI 0x78e1a0e8 ignoring TargetPortalGroupTag=1
iSCSI 0x78e1a0e8 handling AuthMethod=None
iSCSI 0x78e1a0e8 entering operational negotiation
iSCSI 0x78e1a0e8 ignoring HeaderDigest=None
iSCSI 0x78e1a0e8 ignoring DataDigest=None
iSCSI 0x78e1a0e8 ignoring MaxConnections=1
iSCSI 0x78e1a0e8 ignoring InitialR2T=Yes
iSCSI 0x78e1a0e8 ignoring ImmediateData=No
iSCSI 0x78e1a0e8 ignoring MaxBurstLength=262144
iSCSI 0x78e1a0e8 ignoring MaxBurstLength=65536
iSCSI 0x78e1a0e8 ignoring DefaultTime2Wait=0
iSCSI 0x78e1a0e8 ignoring DefaultTime2Retain=0
iSCSI 0x78e1a0e8 ignoring MaxOutstandingR2T=1
iSCSI 0x78e1a0e8 ignoring DataPDUInOrder=Yes
iSCSI 0x78e1a0e8 ignoring DataSequenceInOrder=Yes
iSCSI 0x78e1a0e8 ignoring ErrorRecoveryLevel=0
iSCSI 0x78e1a0e8 ignoring MaxRecvDataSegmentLength=1048576
iSCSI 0x78e1a0e8 entering full feature phase
SCSI 0x78e1a388 waiting for unit to become ready
SCSI 0x78e1a388 tag 18ae0002 status 02 sense 70 key 00 additional 2900
SCSI 0x78e1a388 tag 18ae0002 closed: Input/output error (http://ipxe.org/1d704098)
SCSI 0x78e1a388 not ready: Input/output error (http://ipxe.org/1d704098)
SCSI 0x78e1a388 retrying (retry 1)
SCSI 0x78e1a388 waiting for unit to become ready
NETDEV net0 transmission 0x0 failed: Error 0x2a086089 (http://ipxe.org/2a086089)
SCSI 0x78e1a38 unit is ready
SAN 0x80.0 is active
NETDEV net0 transmission 0x0 failed: Error 0x2a086089 (http://ipxe.org/2a086089)
NETDEV net0 transmission 0x0 failed: Error 0x2a086089 (http://ipxe.org/2a086089)
SAN 0x80 registered
EFIBLK 0x80 has device path /VenHw(<REDACTED_GUID>)/MAC(<REDACTED_MAC_ADDRESS>,1)/MAC(<REDACTED_MAC_ADDRESS>,1)/UNKNOWN(0003,0013)
NETDEV net0 transmission 0x0 failed: Error 0x2a086089 (http://ipxe.org/2a086089)
NETDEV net0 transmission 0x0 failed: Error 0x2a086089 (http://ipxe.org/2a086089)
NETDEV net0 transmission 0x0 failed: Error 0x2a086089 (http://ipxe.org/2a086089)
iPXE> sanboot --no-describe --filename Image
Booting Image from SAN device 0x80
EFIBLK 0x80 could not connect drivers: Error 0x7f22208e (http://ipxe.org/7f22208e)
EFIBLK 0x80 found child device /VenHw(<REDACTED_GUID>)/MAC(<REDACTED_MAC_ADDRESS>,1)/MAC(<REDACTED_MAC_ADDRESS>,1)/UNKNOWN(0003,0013)/HD(1,MBR,<REDACTED>,0x800,0x20000)
EFIBLK 0x80 trying to load /VenHw(<REDACTED_GUID>)/MAC(<REDACTED_MAC_ADDRESS>,1)/MAC(<REDACTED_MAC_ADDRESS>,1)/UNKNOWN(0003,0013)/HD(1,MBR,<REDACTED>,0x800,0x20000)/Image
NETDEV net0 transmission 0x0 failed: Error 0x2a086089 (http://ipxe.org/2a086089)
EFIBLK 0x80 could not load image: Error 0x7f22208e (http://ipxe.org/7f22208e)
EFIBLK 0x80 found child device /VenHw(<REDACTED>)/MAC(<REDACTED_MAC_ADDRESS>,1)/MAC(<REDACTED_MAC_ADDRESS>,1)/UNKNOWN(0003,0013)/HD(1,MBR,<REDACTED>,0x20800,0x3a365830)
EFIBLK 0x80 trying to load /VenHw(<REDACTED_GUID>)/MAC(<REDACTED_MAC_ADDRESS>,1)/MAC(<REDACTED_MAC_ADDRESS>,1)/UNKNOWN(0003,0013)/HD(1,MBR,<REDACTED>,0x20800,0x3a365830)/Image
NETDEV net0 transmission 0x0 failed: Error 0x2a086089 (http://ipxe.org/2a086089)
EFIBLK 0x80 could not load image: Error 0x7f22208e (http://ipxe.org/7f22208e)
Boot from SAN device 0x80 failed: Error 0x7f22208e (http://ipxe.org/7f22208e)
```



And these are the log prints i get on my centos machine which hosts the iSCSI targets (using SCST):
```
Aug 16 14:41:14: scst: Using security group iqn.2021-07.com.test:target1 for initiator iqn.2021-07.ipxe:test-unit-1.test

Aug 16 14:41:14: iscsi-scst: Session ffff8803d1c7c040 created: target ffff88035d378a00, tid 1, sid 0x1f5f323270040, initiator iqn.2021-07.ipxe:test-unit-1.test

Aug 16 14:41:14: iscsi-scst: Negotiated parameters: InitialR2T Yes, ImmediateData No, MaxConnections 1, MaxRecvDataSegmentLength 1048576, MaxXmitDataSegmentLength 8192 MaxBurstLength 262144, FirstBurstLength 65536, DefaultTime2Wait 0, DefaultTime2Retain 0,MaxOutstandingR2T 1, DataPDUInOrder Yes, DataSequenceInOrder Yes, ErrorRecoveryLevel 0, HeaderDigest None, DataDigest None, OFMarker No, IFMarker No, OFMarkInt 2048, IFMarkInt 2048, RDMAExtension No

Aug 16 14:41:14: iscsi-scst: Target parameters for session 1f5f323270040: QueuedCommands 32, Response timeout 90, Nop-In interval 30, Nop-In timeout 30

Aug 16 14:41:14: iscsi-scst: Creating connection ffff8801a5cd3a80 for sid 0x1f5f323270040, cid 0 (initiator iqn.2021-07.ipxe:test-unit-1.test)

Aug 16 14:41:53: iscsi-scst: ***ERROR***: Timeout 30 sec sending data/waiting for reply to/from initiator initiator iqn.2021-07.ipxe:test-unit-1.test (SID 1f5f323270040), closing connection ffff8801a5cd3a80

Aug 16 14:41:53: scst: TM fn NEXUS_LOSS_SESS/6 (mcmd ffff8802f56da0b8, initiator iqn.2021-07.ipxe:test-unit-1.test, target iqn.2021-07.com.test:target1)

Aug 16 14:41:53: scst: TM fn 6 (mcmd ffff8802f56da0b8) finished, status 0

Aug 16 14:41:53: iscsi-scst: Freeing conn ffff8801a5cd3a80 (sess=ffff8803d1c7c040, 0x1f5f323270040 0, initiator iqn.2021-07.ipxe:test-unit-1.test)

Aug 16 14:41:53: iscsi-scst: Freeing session ffff8803d1c7c040 (SID 1f5f323270040)
```

**Reply by NiKiZe** [reply]
Thanks for verifying using recent version of iPXE
Please read this one, http://ipxe.org/1d704098 - I would like to think that it would be easier to spot without the debug prints. (It is very common to see missing or incorrect option for LUN)

Do a test with only `sanboot`, no `sanhook`

Note that without working `iBFT` (which you disable with `--no-describe`) you will need to have the iSCSI configuration in the OS.
This is a firmware issue, not iPXE.

**Reply by alan-alanson** [reply]
I have managed to boot the kernel using sanboot. Unfortunately, I do not think my firmware (uBoot) supports iBFT for arm64 currently :( , meaning I have to use `--no-describe`
My current solution is to use an initramfs that will log in to the iSCSI target and mount the root file system and continue on with boot. This works as I have built the initramfs in to the kernel image.

However, I am not sure what to do about the device tree. My linux device tree is distinct from the one my firmware (uBoot) uses. Therefore I ideally need a way of supplying it to the kernel when it gets executed. [Here ](https://www.kernel.org/doc/html/latest/admin-guide/efi-stub.html) suggests that I can use a dtb= parameter when executing the kernel's efi stub, however I do not believe passing parameters such as this are supported when using sanboot ?

**Reply by NiKiZe** [reply]
If you use `sanboot` then you will have a "disk" and all data need to be on the disk or have other ways to propagate, such as ACPI tables.
But it does sound like the best approach for you is to use kernel + initramfs and add parameters on the cmdline, one such parameter could be iscsi root.

You will not have iSCSI boot all the way but you will have iSCSI root, is that not an option for some reason?

**Reply by alan-alanson** [reply]
iSCSI root fs was a requirement, iSCSI boot all the way was the dream! Looks like I will have to settle for the former.

Thanks a lot for your help!

**Reply by mcb30** [reply]
> I have managed to boot the kernel using sanboot. Unfortunately, I do not think my firmware (uBoot) supports iBFT for arm64 currently :( , meaning I have to use `--no-describe`

The iBFT is an ACPI table.  If (as suggested above) your firmware includes no support for ACPI then it will not include the mechanism that iPXE would need to use to register a new ACPI table, and so you would unfortunately not be able to rely on the iBFT to hand over iSCSI information to the booted kernel+initramfs.

> However, I am not sure what to do about the device tree. My linux device tree is distinct from the one my firmware (uBoot) uses. Therefore I ideally need a way of supplying it to the kernel when it gets executed. [Here ](https://www.kernel.org/doc/html/latest/admin-guide/efi-stub.html) suggests that I can use a dtb= parameter when executing the kernel's efi stub, however I do not believe passing parameters such as this are supported when using sanboot ?

No, there is no way to pass command-line parameters to a kernel from a SAN-booted disk, since the operating model is "boot this disk" rather than "boot this executable program".
